src/analysis/normed_space/hahn_banach.lean

```diff
@@ -95,48 +124,53 @@ begin
   { exact f.op_norm_le_bound g.extend_to_ℂ.op_norm_nonneg (λ x, h x ▸ g.extend_to_ℂ.le_op_norm x) },
 end
 
+instance complex_has_exists_extension_norm_eq : has_exists_extension_norm_eq ℂ :=
+⟨by { intros, apply complex.exists_extension_norm_eq }⟩
+
 end complex
 
-section dual_vector
-variables {E : Type*} [normed_group E] [normed_space ℝ E]
+variables {𝕜 : Type u} [nondiscrete_normed_field 𝕜] [normed_algebra ℝ 𝕜]
+variables {E : Type v} [normed_group E] [normed_space 𝕜 E]
 
 open continuous_linear_equiv
 open_locale classical
 
-lemma coord_self' (x : E) (h : x ≠ 0) : (∥x∥ • (coord ℝ x h))
-  ⟨x, submodule.mem_span_singleton_self x⟩ = ∥x∥ :=
-calc (∥x∥ • (coord ℝ x h)) ⟨x, submodule.mem_span_singleton_self x⟩
-    = ∥x∥ • (linear_equiv.coord ℝ E x h) ⟨x, submodule.mem_span_singleton_self x⟩ : rfl
-... = ∥x∥ • 1 : by rw linear_equiv.coord_self ℝ E x h
-... = ∥x∥ : mul_one _
+lemma coord_self' (x : E) (h : x ≠ 0) : (coe_from_ℝ 𝕜 (∥x∥) • (coord 𝕜 x h))
+  ⟨x, submodule.mem_span_singleton_self x⟩ = coe_from_ℝ 𝕜 (∥x∥) :=
+calc ((coe_from_ℝ 𝕜 (∥x∥)) • (coord 𝕜 x h)) ⟨x, submodule.mem_span_singleton_self x⟩
+    = (coe_from_ℝ 𝕜 (∥x∥)) • (linear_equiv.coord 𝕜 E x h) ⟨x, submodule.mem_span_singleton_self x⟩ : rfl
+... = (coe_from_ℝ 𝕜 (∥x∥)) • 1 : by rw linear_equiv.coord_self 𝕜 E x h
+... = (coe_from_ℝ 𝕜 (∥x∥)) : mul_one _
+
+lemma coord_norm' (x : E) (h : x ≠ 0) : ∥(coe_from_ℝ 𝕜 (∥x∥)) • coord 𝕜 x h∥ = 1 :=
+by rw [norm_smul, norm_norm', coord_norm, mul_inv_cancel (mt norm_eq_zero.mp h)]
 
-lemma coord_norm' (x : E) (h : x ≠ 0) : ∥∥x∥ • coord ℝ x h∥ = 1 :=
-by rw [norm_smul, norm_norm, coord_norm, mul_inv_cancel (mt norm_eq_zero.mp h)]
+variables [has_exists_extension_norm_eq.{u v} 𝕜]
 
 /-- Corollary of Hahn-Banach.  Given a nonzero element `x` of a normed space, there exists an
     element of the dual space, of norm 1, whose value on `x` is `∥x∥`. -/
-theorem exists_dual_vector (x : E) (h : x ≠ 0) : ∃ g : E →L[ℝ] ℝ, ∥g∥ = 1 ∧ g x = ∥x∥ :=
+theorem exists_dual_vector (x : E) (h : x ≠ 0) : ∃ g : E →L[𝕜] 𝕜, ∥g∥ = 1 ∧ g x = (coe_from_ℝ 𝕜 (∥x∥)) :=
 begin
-  cases exists_extension_norm_eq (submodule.span ℝ {x}) (∥x∥ • coord ℝ x h) with g hg,
+  let p : submodule 𝕜 E := submodule.span 𝕜 {x},
+  let f := (coe_from_ℝ 𝕜 ∥x∥) • coord 𝕜 x h,
+  obtain ⟨g, hg⟩ := has_exists_extension_norm_eq.exists_extension_norm_eq E p f,
   use g, split,
   { rw [hg.2, coord_norm'] },
-  { calc g x = g (⟨x, submodule.mem_span_singleton_self x⟩ : submodule.span ℝ {x}) : by simp
-  ... = (∥x∥ • coord ℝ x h) (⟨x, submodule.mem_span_singleton_self x⟩ : submodule.span ℝ {x}) : by rw ← hg.1
-  ... = ∥x∥ : by rw coord_self' }
+  { calc g x = g (⟨x, submodule.mem_span_singleton_self x⟩ : submodule.span 𝕜 {x}) : by rw submodule.coe_mk
+  ... = ((coe_from_ℝ 𝕜 (∥x∥)) • (coord 𝕜 x h)) (⟨x, submodule.mem_span_singleton_self x⟩ : submodule.span 𝕜 {x}) : by rw ← hg.1
+  ... = (coe_from_ℝ 𝕜 (∥x∥)) : by rw coord_self' }
```

Comment:
These lines are quite long. Perhaps `open submodule` in this section so we don't need to repeat `submodule` everywhere?